File: packages/agent/src/infrastructure/events/types/memory.ts
/**
 * @fileoverview Memory Events
 *
 * Events for the memory ledger system. Ledger entries are structured
 * summaries of response cycles, written by a background Haiku subagent.
 */

import type { BaseEvent } from './base.js';

// =============================================================================
// Memory Ledger Payload
// =============================================================================

export interface MemoryLedgerPayload {
  /** Event range this entry covers */
  eventRange: { firstEventId: string; lastEventId: string };
  /** Turn range this entry covers */
  turnRange: { firstTurn: number; lastTurn: number };
  /** Short title for the entry */
  title: string;
  /** Classification of the work done */
  entryType: 'feature' | 'bugfix' | 'refactor' | 'docs' | 'config' | 'research' | 'conversation';
  /** Current status */
  status: 'completed' | 'partial' | 'in_progress';
  /** Tags for categorization */
  tags: string[];
  /** Original user request summary */
  input: string;
  /** Actions taken */
  actions: string[];
  /** Files touched */
  files: Array<{ path: string; op: 'C' | 'M' | 'D'; why: string }>;
  /** Key decisions and rationale */
  decisions: Array<{ choice: string; reason: string }>;
  /** Patterns and lessons for future reference */
  lessons: string[];
  /** Key reasoning insights from thinking blocks */
  thinkingInsights: string[];
  /** Token cost of the cycle */
  tokenCost: { input: number; output: number };
  /** Model used */
  model: string;
  /** Working directory at time of entry */
  workingDirectory: string;
}

// =============================================================================
// Memory Events
// =============================================================================

export interface MemoryLedgerEvent extends BaseEvent {
  type: 'memory.ledger';
  payload: MemoryLedgerPayload;
}

File: packages/agent/crates/tron-events/src/types/payloads/memory.rs
//! Memory event payloads.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Payload for `memory.ledger` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryLedgerPayload {
    /// Event range covered.
    pub event_range: EventRange,
    /// Turn range covered.
    pub turn_range: TurnRange,
    /// Session title.
    pub title: String,
    /// Entry type.
    pub entry_type: String,
    /// Completion status.
    pub status: String,
    /// Tags.
    pub tags: Vec<String>,
    /// Original user input/request.
    pub input: String,
    /// Actions taken.
    pub actions: Vec<String>,
    /// Files modified.
    pub files: Vec<LedgerFileEntry>,
    /// Decisions made.
    pub decisions: Vec<LedgerDecision>,
    /// Lessons learned.
    pub lessons: Vec<String>,
    /// Insights from thinking blocks.
    pub thinking_insights: Vec<String>,
    /// Token costs.
    pub token_cost: LedgerTokenCost,
    /// Model used.
    pub model: String,
    /// Working directory.
    pub working_directory: String,
}

/// Event ID range.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EventRange {
    /// First event ID.
    pub first_event_id: String,
    /// Last event ID.
    pub last_event_id: String,
}

/// Turn number range.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnRange {
    /// First turn number.
    pub first_turn: i64,
    /// Last turn number.
    pub last_turn: i64,
}

/// File entry in a ledger record.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct LedgerFileEntry {
    /// File path.
    pub path: String,
    /// Operation: C (create), M (modify), D (delete).
    pub op: String,
    /// Purpose description.
    pub why: String,
}

/// Decision in a ledger record.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct LedgerDecision {
    /// What was chosen.
    pub choice: String,
    /// Why it was chosen.
    pub reason: String,
}

/// Token cost in a ledger record.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct LedgerTokenCost {
    /// Input tokens.
    pub input: i64,
    /// Output tokens.
    pub output: i64,
}

/// Payload for `memory.loaded` events.
///
/// Kept as opaque JSON since the schema is flexible.
pub type MemoryLoadedPayload = Value;
